server/src/test/java/org/elasticsearch/env/NodeEnvironmentIT.java

```diff
@@ -86,4 +97,44 @@ public Settings onNodeStopped(String nodeName) {
                 + Node.NODE_DATA_SETTING.getKey()
                 + "=false, but has shard data"));
     }
+
+    private IllegalStateException expectThrowsOnRestart(CheckedConsumer<Path[], Exception> onNodeStopped) {
+        internalCluster().startNode();
+        final Path[] dataPaths = internalCluster().getInstance(NodeEnvironment.class).nodeDataPaths();
+        return expectThrows(IllegalStateException.class,
+            () -> internalCluster().restartRandomDataNode(new InternalTestCluster.RestartCallback() {
+                @Override
+                public Settings onNodeStopped(String nodeName) {
+                    try {
+                        for (Path dataPath : dataPaths) {
```

Comment:
This loop seems to be removing node metadata files. Why is it needed? You anyway re-write the files in the tests.

Reply:
Ah, this is a throwback to a version of this PR that failed the node if the node metadata file was removed but the data directory was nonempty. I think we should be stricter about starting a node on such a data directory, but that change has wider impact on the tests and is worthy of a separate discussion. Removed in 3320d5a75abc24f3b389b12c2c807f83152ec19c.